cmd/gimlet/src/lib.rs

```diff
@@ -0,0 +1,151 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! ## `humility gimlet`
+//!
+//! `humility gimlet` contacts a (recent firmware version) Gimlet over the
+//! management network and extracts certain live diagnostic information.
+//!
+//! ```console
+//! $ humility gimlet --ip fe80::0c1d:9aff:fe64:b8c2%en0 read-seq-regs
+//! ```
+//!
+//! For a complete list of subcommands, use `humility gimlet --help`.
+
+use std::net::{SocketAddrV6, UdpSocket};
+use std::time::Duration;
+
+use anyhow::{bail, Context, Result};
+use clap::{ArgGroup, IntoApp, Parser};
+use humility::net::decode_iface;
+use humility_cli::{ExecutionContext, Subcommand};
+use humility_cmd::{Archive, Command, CommandKind, Dumper};
+
+/// This is defined in the gimlet TOML.
+const HARDCODED_PORT: u16 = 23547;
+
+#[derive(Parser, Debug)]
+#[clap(
+    name = "gimlet", about = env!("CARGO_PKG_DESCRIPTION"),
+    group = ArgGroup::new("target").multiple(false)
+)]
+struct Args {
+    /// How long to wait for a response from the target, in milliseconds.
+    #[clap(
+        long, short = 'T', default_value_t = 2000, value_name = "ms",
+        parse(try_from_str = parse_int::parse)
+    )]
+    timeout: u64,
+
+    /// UDP port to contact on the target.
+    #[clap(long, default_value_t = HARDCODED_PORT)]
+    port: u16,
+
+    /// IPv6 address, e.g. `fe80::0c1d:9aff:fe64:b8c2%en0`
+    #[clap(
+        long,
+        env = "HUMILITY_RPC_IP",
+        group = "target",
+        use_value_delimiter = true
+    )]
+    ip: String,
```

Comment:
I was going to mention that  [`std::net::Ipv6Addr` implements `FromStr`](https://doc.rust-lang.org/stable/std/net/struct.Ipv6Addr.html#impl-FromStr-for-Ipv6Addr), so this could be changed from `String` to `Ipv6Addr` and `clap` would print a nicer error if it can't be parsed. But, I see that we do some additional parsing of the string to extract a scope, so that would be a bit more complex.

We could, potentially, add a new type like
```rust
struct ScopedV6Addr {
   scope: u32,
   addr: Ipv6Addr,
}

impl FromStr for ScopedV6Addr {
    // ...
}
```

but, this might be a bit more effort and it's quite reasonable if you feel like it's out of scope (hah!) for this PR...

Reply:
I might just go and throw together a separate branch making this change, actually.

Reply:
The scope-parsing code is now duplicated across probably a half-dozen locations in different repos, so centralizing it would be awesome. Rust std doesn't parse scopes in a useful manner -- it only parses integers, which is a narrow subset of the RFC -- so we wind up needing to do it ourselves.

Reply:
@cbiffle I've opened  #443 to make this change in Humility. In the longer term, I could go ahead and make a (tiny) crate for it, so that we can use it in  other repos as well.

Reply:
Neat! I see you picked up the `rsplit_once` simplification too. I'll take a look!

Reply:
Yeah, I liked that --- although I went back and forth on it for a sec, because I thought that in the case where the IP is invalid _and_ there's no `%scope`, it might be nicer to error on "hey, 10.0.01 is not an IPv6 address!" than "also you forgot %eth0", but :woman_shrugging: --- i don't know if that actually matters.